edb/lib/ext/pgvector.edgeql

```diff
@@ -0,0 +1,179 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2023-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+create extension package pgvector version '0.4.0' {
+    set ext_module := "ext::pgvector";
+    set sql_extensions := ["vector >=0.4.0,<0.5.0"];
+
+    create module ext::pgvector;
+
+    create scalar type ext::pgvector::vector extending std::anyscalar {
+        set id := <uuid>"9565dd88-04f5-11ee-a691-0b6ebe179825";
+        set sql_type := "vector";
+        set sql_type_scheme := "vector({__arg_0__})";
+        set num_params := 1;
+    };
+
+    create cast from ext::pgvector::vector to std::json {
+        set volatility := 'Immutable';
+        using sql 'SELECT val::text::jsonb';
+    };
+
+    create cast from std::json to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT (
+            nullif(val, 'null'::jsonb)::text::vector
+        )
+        $$;
+    };
+
+    # All casts from numerical arrays should allow assignment casts.
+    create cast from array<std::float32> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::float64> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::int16> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT val::float4[]::vector
+        $$;
+        allow assignment;
+    };
+
+    create cast from array<std::int32> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::int64> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT val::float4[]::vector
+        $$;
+        allow assignment;
+    };
+
+    create cast from array<std::decimal> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql cast;
+        allow assignment;
+    };
+
+    create cast from array<std::bigint> to ext::pgvector::vector {
+        set volatility := 'Immutable';
+        using sql $$
+        SELECT val::numeric[]::vector
+        $$;
+        allow assignment;
+    };
+
```

Comment:
I don't think we need decimal or bigint/decimal casts -- we don't need them and it's questionable if they're useful at all, as there will be major loss of precision here

Reply:
We allow `float <> decimal/bigint` casts, though.

Reply:
E.g

```
select <array<float32>>[1n, 2n, 3n];
```

works, and I don't see a fundamental difference between `array<float32>` and `pgvector::vector` in terms of casting.